### PR TITLE
aws-prune replacement 

### PR DESCRIPTION
**Pull request by oktopuz**
To solve the issue #43 

Delete all terminated or missing AWS clients from sensu

1. Get Sensu API creds from /etc/sensu/conf.d/api.json
2. Get Sensu clients that have a valid instance_id
3. Get AWS instance_id list from /var/cache/instance_list.json
4. Find a diff between two lists
5. Delete all sensu clients listed in diff

### CODE REVIEW

**Comment by somic** [comment]
How are we going to call this script? Could do it from cron on each sensu server but sounds wasteful (plus race conditions when deleting clients from sensu, even though in case of deletion it's likely safe). Or could do it as a server side check (monitoring_check::server_side).

Another concern I have is freshness of /var/cache/instance_list.json and the fact that we are using it as the single source of truth. We may end up in a situation when this file on this host is not up to date, for whatever reason. In that case, this script could be deleting valid hosts.

Current version of aws_prune has a built in protection against this because it's implemented as a handler and is called only when sensu triggers a keepalive event for a client.

I can't think of a good solution except for manually checking modification timestamp on a file, which I can't say I like very much. Or get a list of instances from AWS ourselves here and not use cached version? Why have a cache then if we are not using it?... :smile:

**Comment by solarkennedy** [comment]
I'm going to just respond to @somic  and do general review in one post:

Yes. This script should not use the cached instance list, it should just get the list from the api. We only needed the cache before because of all the calls, the very reason for doing it this way is so we don't need to be so aggressive as to need the complexity of a cache.

Additionally, @oktopuz can you greatly simplify this code by not requiring it to be "sensu-aware" at all and reusing the sensu-cli tool instead? Then you could something like:

    sensu-cli client list | list_stale_ec2_servers | xargs -x -n1 sensu-client delete

I think that would be pretty cool. You could take the unix'ness to the extreme, but I think this is an ok compromise.

I think it would be fine to just run this on every sensu server once a minute.

**Comment by oktopuz** [comment]
@solarkennedy even though initially I liked the idea of using sensu-cli in the beginning, I don't think it is a safe solution, because we can't control what is sent to sensu-cli for deletion.
For instance, if you run cache_instance_list.rb now, you'll see a lot of data sent to STDOUT (!) because of some issue in the fog gem. Who knows what will happen if this output goes directly to 'sensu-cli client delete'?

I don't see any way to guarantee that list_stale_ec2_servers will send valid output only. That is why I believe that API-based solution would be the good way to go.

Right now I am replacing fog related functionality by aws-sdk v2 to query AWS API directly. Meanwhile we can decide whether to use Sensu API or unix pipes.

**Comment by solarkennedy** [comment]
Ok.

File: files/sensu_cleanup_aws_clients.rb [hunk 1]
```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env ruby
```
Comment (solarkennedy) [inline files/sensu_cleanup_aws_clients.rb]: Although this is an open source project, you won't be able to use the system ruby (probably)

Our puppet ruby has the aws sdk. Alternatively you could install the aws-sdk using sensu's omnibus ruby (following the pattern of the other handlers)

File: files/sensu_cleanup_aws_clients.rb [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+    if !response.nil?
+      log_delete_result(client, response.code)
+    end
+  end
+
+  def log_delete_result(client, code)
+    case code
+    when '202'
+        @logger.info("EC2 Node - [202] Successfully deleted Sensu client: #{client}")
```
Comment (solarkennedy) [inline files/sensu_cleanup_aws_clients.rb]: This logging should change, you don't have to say "EC2 Node - "

**Comment by solarkennedy** [comment]
I think this is a good start. I don't really like that it is a lot of code and no tests at all.

I encourage you to look at the trollop gem to help simplify the arg parsing.

It isn't like the previous code for this was super great (I wrote it), but I'm ok with you proceeding as is, shipping this *in parallel* (noop mode), testing and iterating till it can replace the existing handler.

**Comment by oktopuz** [comment]
I've added rspec tests: https://github.com/oktopuz/sensu_handlers/commit/1f4b0ea840e1fd6022cb146c45cd64a9bcb15a23
Going to take a look at trollop.

File: files/delete_terminated_ec2_clients.rb [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+      @logger = logger
+    end
+
+    Aws.config.update({
+      :region      => region,
+      :credentials => Aws::Credentials.new(aws_access_key_id, aws_secret_access_key),
+    })
+  end
+
+  def get_ec2_instanses_info
```
Comment (hashbrowncipher) [inline files/delete_terminated_ec2_clients.rb]: Please spell this 'instances'

File: files/delete_terminated_ec2_clients.rb [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    end
+
+    aws_instances = @aws.get_ec2_instanses_info.reject { |id, val| val['state'] == 'terminated' }
+    @logger.debug("Found #{aws_instances.keys.count} non-terminated AWS instances.")
+
+    diff = sensu_clients.keys - aws_instances.keys
+    hosts_to_delete = []
+    diff.each { |id| hosts_to_delete << sensu_clients[id] }
+
+    @logger.info(diff.count > 0 ? "#{diff.count} Sunsu clients to delete: " +
```
Comment (hashbrowncipher) [inline files/delete_terminated_ec2_clients.rb]: Sensu

**Comment by oktopuz** [comment]
This is complete.
Could you review it please?

**Comment by solarkennedy** [comment]
Does this need more review? I thought it was already in production?

**Comment by oktopuz** [comment]
Yes, it is in production.
I believe it is safe to merge.

**Comment by solarkennedy** [comment]
But if this is in production, why the pull request? Are we not running master?

**Comment by oktopuz** [comment]
It is now a part of our internal sensu_server_side class.
The only reason for this merge I can think of is to make it available as a part of sensu_handlers.

**Comment by solarkennedy** [comment]
Ah. If you ask me, that is the "right" place to put it. It isn't a handler, so it doesn't need to be in the sensu_handlers project.

**Comment by oktopuz** [comment]
Agree. I am closing the PR.

**Pull request closed without merging**